src/ngraph/function.cpp

```diff
@@ -224,3 +224,19 @@ void Function::set_placement(size_t placement)
 {
     m_placement = placement;
 }
+
+// TODO(pthoreho) this will be expensive, since we will be traversing all the nodes in
+// the graph, figure out if their is a way to cache the result and invalidate/update
+// the result if the function is modified
+bool Function::is_dynamic() const
```

Comment:
Let's call this `update_dynamic` and make it non const, so it can cache the result in `Function`. Let `is_dynamic` return the cached value. We can have the `Function` constructor call `update_dynamic` and we can call it from passes if we do things that might make a function non-dynamic through an optimization.

Reply:
@diyessi I think in most situations any function that takes a `ngraph::Function` object as a parameter can not make the assumption that `is_dynamic()` has the most up to date information and will need to always call `update_dynamic()` then call `is_dynamic()`. So even if a pass that changes a `Function` object state calls `update_dynamic`, another pass that take that resulting `Function` still have to call `update_dynamic()` before it can be certain that `is_dynamic()" returns correct information. 

Giving `Function` the responsibility to cache the value but does not control the logic when to update and keep that state consistent can be error prone. I would recommend keeping it simple right now with `is_dynamic()` always return the correct result and let the user determine when to cache the result.

Reply:
@diyessi i also agree with @louisfeng , we thought about caching the `is_dynamic` result. but since function state can be changed outside the graph pass. We thought of keeping the API simple for now to keep us going.  this approach even though have cost associated with it but less error prone for now i believe.